z3/src/datatype_builder.rs

```diff
@@ -1,117 +1,179 @@
+    Symbol,
+};
 
 impl<'ctx> DatatypeBuilder<'ctx> {
-    pub fn new(ctx: &'ctx Context) -> Self {
+    pub fn new<S: Into<Symbol>>(ctx: &'ctx Context, name: S) -> Self {
         Self {
             ctx,
-            variants: Vec::new(),
+            name: name.into(),
+            constructors: Vec::new(),
         }
     }
 
-    pub fn variant(mut self, name: &str, fields: &[(&str, &Sort)]) -> Self {
-        let recognizer_name_sym = Symbol::String(format!("is-{}", name)).as_z3_symbol(self.ctx);
-        let name_sym = Symbol::String(name.to_string()).as_z3_symbol(self.ctx);
-
-        assert!(fields
-            .iter()
-            .all(|(name, sort)| sort.ctx.z3_ctx == self.ctx.z3_ctx));
-
-        let mut field_names: Vec<Z3_symbol> = Vec::with_capacity(fields.len());
-        let mut field_sorts = Vec::with_capacity(fields.len());
-
-        for (name, sort) in fields {
-            field_names.push(Symbol::String(name.to_string()).as_z3_symbol(self.ctx));
-            field_sorts.push(sort.z3_sort);
+    pub fn variant(&mut self, name: &str, fields: &'ctx [(&str, DatatypeAccessor<'ctx>)]) {
+        let mut accessor_vec: Vec<(String, &DatatypeAccessor<'ctx>)> = Vec::new();
+        for (accessor_name, accessor) in fields {
+            accessor_vec.push((accessor_name.to_string(), accessor));
         }
 
-        // This is unused.
-        // Z3 expects sort_refs in Z3_mk_constructor to be valid, so create it here.
-        let mut sort_refs = Vec::new();
-        sort_refs.resize(fields.len(), 0);
-
-        let constructor = unsafe {
-            Z3_mk_constructor(
-                self.ctx.z3_ctx,
-                name_sym,
-                recognizer_name_sym,
-                fields.len().try_into().unwrap(),
-                field_names.as_ptr(),
-                field_sorts.as_ptr(),
-                sort_refs.as_mut_ptr(),
-            )
-        };
+        self.constructors.push((name.to_string(), accessor_vec));
+    }
 
-        self.variants.push((fields.len(), constructor));
-        self
+    pub fn finish(&self) -> DatatypeSort<'ctx> {
+        let mut dtypes = create_datatypes(&[self]);
+        dtypes.remove(0)
     }
+}
 
-    pub fn finish(self, name: impl Into<Symbol>) -> DatatypeSort<'ctx> {
-        let mut constructors: Vec<_> = self.variants.iter().map(|i| i.1).collect();
-        let name_sym = name.into().as_z3_symbol(self.ctx);
-
-        let sort = unsafe {
-            let s = Z3_mk_datatype(
-                self.ctx.z3_ctx,
-                name_sym,
-                constructors.len().try_into().unwrap(),
-                constructors.as_mut_ptr(),
-            );
-            Z3_inc_ref(self.ctx.z3_ctx, Z3_sort_to_ast(self.ctx.z3_ctx, s));
-            Sort {
-                ctx: self.ctx,
-                z3_sort: s,
+pub fn create_datatypes<'ctx>(ds: &[&DatatypeBuilder<'ctx>]) -> Vec<DatatypeSort<'ctx>> {
```

Comment:
Does this need to be public? I guess if you have multiple data types that each reference each other?

Let's rescope this to a static method of `DatatypeBuilder`: `DatatypeBuilder::create_datatypes`.

Lets also add a smoke test for mutually recursive data types (which was the whole point of this PR, right?).

Reply:
Yes, to supoprt mutually recursive datatypes. 

Self-recursive datatype test is in tests/lib.rs: https://github.com/prove-rs/z3.rs/blob/9ef0ad8a6b2b45a567d7a8628ac5346d2b508232/z3/tests/lib.rs#L577